### PR TITLE
Support v4 point estimate `required` null values

### LINKED ISSUE #156: Add support for v4 point estimate `NULL` output type_ids to `expand_model_out_grid()`
Currently returns a zero row and column dataframe for point estimate output types

### PR DESCRIPTION
**Pull request by annakrystalli**
This PR resolves #156 by adding support for `NULL` `required` properties in v4 config for point estimates by effectively converting them to `NA`s in `expand_model_out_grid()`

Overall we're converting v4 configuration to pre v4 configurations which means we leverage previous functionality and ensure back-compatibility. Give task IDs also still use both `required` and `optional` properties, this approach means we handle config values consistently.

I've also tried to start addressing #147 in this PR too.

Just a heads up too that, in the next PR (still in prep), I'm actually re-factoring a little of this as I move on to address #159

### CODE REVIEW

**Review by zkamvar — CHANGES_REQUESTED** [review]
This looks good to me. The catch for v4 point estimates makes sense. I have a few suggestions for readability. 

1. use `round` instead of `rnd`. My brain always wants to convert `rnd` to `random` and we don't currently use `rnd` anywhere.
2. use `setequal()` instead of `isTRUE(all.equal(sort())`.

I also wanted to say that I appreciate the explanatory comments and splitting things into functions. This makes things much easier to read!

File: R/expand_model_out_grid.R [hunk 1]
```diff
@@ -194,25 +194,10 @@ expand_model_out_grid <- function(config_tasks,
...
-      } else {
-        mt_output_types <- output_types[output_types %in% names(out)]
-        out[mt_output_types]
-      }
-    }
-  ) %>%
-    purrr::map(
-      ~ extract_mt_output_type_ids(.x, config_tid)
-    ) %>%
+  output_type_l <- subset_rnd_output_types(round_config, output_types) %>%
```
[the remark this answers is not in the document]
Comment (annakrystalli) [reply R/expand_model_out_grid.R:197]: Sure. Fixed in fef4e3a33527a9df86193bda8b6f88b26ac4bc70

File: R/expand_model_out_grid.R [hunk 2]
```diff
@@ -241,6 +226,29 @@ expand_model_out_grid <- function(config_tasks,
   )
 }
 
+# Subset output types according to `output_types` from all model_task objects in
+# a round. If `output_types` is `NULL`, all output types for each model task are
+# returned.
+subset_rnd_output_types <- function(round_config, output_types) {
```
[the remark this answers is not in the document]
Comment (annakrystalli) [reply R/expand_model_out_grid.R:232]: Sure. Fixed in fef4e3a33527a9df86193bda8b6f88b26ac4bc70

File: R/expand_model_out_grid.R [hunk 3]
```diff
@@ -194,25 +194,10 @@ expand_model_out_grid <- function(config_tasks,
...
-        mt_output_types <- output_types[output_types %in% names(out)]
-        out[mt_output_types]
-      }
-    }
-  ) %>%
-    purrr::map(
-      ~ extract_mt_output_type_ids(.x, config_tid)
-    ) %>%
+  output_type_l <- subset_rnd_output_types(round_config, output_types) %>%
+    extract_rnd_output_type_ids(config_tid) %>%
```
[the remark this answers is not in the document]
Comment (annakrystalli) [reply R/expand_model_out_grid.R:198]: Sure. Fixed in fef4e3a33527a9df86193bda8b6f88b26ac4bc70

File: tests/testthat/test-expand_model_out_grid.R [hunk 4]
```diff
@@ -551,5 +551,21 @@ test_that("(#123) expand_output_type_grid() returns expected outputs with option
   )
   expect_equal(nrow(i_have_eight_rows), 8)
   expect_equal(ncol(i_have_eight_rows), 3)
+})
+
+test_that("v4 point estimate output type IDs extracted correctly as NAs", {
+  hub_path <- system.file("testhubs", "v4", "flusight", package = "hubUtils")
+  file_name <- "hub-baseline/2023-05-01-hub-baseline.csv"
+  round_id <- parse_file_name(file_name)$round_id
+  config_tasks <- suppressWarnings(read_config(hub_path = hub_path))
```
[the remark this answers is not in the document]
Comment (annakrystalli) [reply tests/testthat/test-expand_model_out_grid.R:560]: Done in 9fd5fedec06362aee85404c95d3dd2679f1b5145

File: tests/testthat/test-expand_model_out_grid.R [hunk 5]
```diff
@@ -551,5 +551,21 @@ test_that("(#123) expand_output_type_grid() returns expected outputs with option
...
 
+  expect_true(
+    expand_model_out_grid(
+      config_tasks = config_tasks,
+      round_id = round_id,
+      output_types = "mean",
+    )[["output_type_id"]] |>
+      is.na() |>
+      all()
+  )
```
[the remark this answers is not in the document]
Comment (annakrystalli) [reply tests/testthat/test-expand_model_out_grid.R:570]: Nice. Added here: ab83e0a8b2dae1a428d2f467ab33e522fd616266

File: R/expand_model_out_grid.R [hunk 6]
```diff
@@ -538,25 +546,71 @@ get_sample_n <- function(x, config_tid) {
...
+valid_output_type_ids <- function(output_type_ids) {
+  # If output_type_id values are provided and when dealing with an older
+  # config version that has "required" and"optional" fields or  extract output_type_id values
+  has_output_type_ids <- !is.null(output_type_ids)
+  pre_v4 <- isTRUE(
+    all.equal(
+      sort(names(output_type_ids)),
+      sort(c("required", "optional"))
+    )
+  )
```
[the remark this answers is not in the document]
Comment (annakrystalli) [reply R/expand_model_out_grid.R:592]: Thanks! much better. Fixed in 64796e0e9f7ff2dc38796b027ced21e44544b2b9

[the remark this answers is not in the document]
Comment (annakrystalli) [reply R/expand_model_out_grid.R:592]: Note I had already pulled it out into a function in the next PR

**Comment by annakrystalli** [comment]
Thanks for the thorough review @zkamvar ! I know it's not ideal but because I'd moved on and done quite a bit of refactoring already in the next PR, I've opted to make the changes in that branch to avoid a messy merge mission.  I've added links to each commit though so you can track the changes. Hope that's ok!

**Review by zkamvar — APPROVED** [review]

**Pull request merged**